### PR TITLE
Feat #1130 [WIP] Adding support for downloading and zipping files

### PR DESCRIPTION
**Pull request by emfol**
Introducing a new extension to support downloading and zipping SOP instances from a study

### PR Checklist

- [x] Brief description of changes
- [ ] Links to any relevant issues
- [ ] Required status checks are passing
- [ ] User cases if changes impact the user's experience
- [ ] `@mention` a maintainer to request a review

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### CODE REVIEW

File: extensions/dicom-p10-downloader/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+# @ohif/extension-dicom-pdf
```
Comment (dannyrb) [inline extensions/dicom-p10-downloader/README.md]: Fix name

File: extensions/dicom-p10-downloader/src/commandsModule.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  getDicomWebClientFromContext,
+  getSOPInstanceReferenceFromActiveViewport,
+} from './utils';
+import downloadAndZip from './downloadAndZip';
+
+export function getCommands(context) {
+  const actions = {
+    downloadAndZipSeriesOnActiveViewport({ servers, viewports, progress }) {
+      const dicomWebClient = getDicomWebClientFromContext(context, servers);
+      if (dicomWebClient) {
```
Comment (swederik) [inline extensions/dicom-p10-downloader/src/commandsModule.js]: Shouldn't we just do if (!dicomWebClient) throw new Error...

Failing silently doesn't seem good here

File: extensions/dicom-p10-downloader/src/commandsModule.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        const reference = getSOPInstanceReferenceFromActiveViewport(viewports);
+        if (reference) {
+          downloadAndZip(dicomWebClient, reference, { progress })
+            .then(url => {
+              OHIF.log.info('Zip file successfully created:', url);
+              location.href = url;
```
Comment (swederik) [inline extensions/dicom-p10-downloader/src/commandsModule.js]: Is this going to redirect the current tab? We don't want that, it might disrupt their work. Better to force a download. This is how we do it with the SaveAs tool in Cornerstone: https://github.com/cornerstonejs/cornerstoneTools/blob/master/src/util/SaveAs.js#L34

File: extensions/dicom-p10-downloader/src/downloadAndZip.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  // Print tree of hierarchical references
+  OHIF.log.info('Downloading DICOM P10 files for references:');
+  OHIF.log.info(hierarchicalListUtils.print(listOfUIDs));
+
+  // Wait on created download promises
+  return Promise.all(downloads).then(results => {
+    const buffers = [];
+    // The "results" array may directly contain buffers (ArrayBuffer instances)
+    // or arrays of buffers, depending on the type of downloads initiated on the
+    // previous step (retrieveStudy, retrieveSeries or retrieveinstance). Ex:
```
Comment (swederik) [inline extensions/dicom-p10-downloader/src/downloadAndZip.js]: This is making me think we should change dicomweb client later to always return an array for consistency.

File: extensions/dicom-p10-downloader/src/index.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+   * Only required property. Should be a unique value across all extensions.
+   */
+  id: 'dicom-p10-downloader',
+
+  /**
+   * LIFECYCLE HOOKS
+   */
+
+  preRegistration({ appConfig }) {
+    const dicomWebClient = getDicomWebClientFromConfig(appConfig);
```
Comment (swederik) [inline extensions/dicom-p10-downloader/src/index.js]: So if the server ever changes, this will stop working? (not that we can currently change servers, just curious)

File: platform/viewer/src/routes/routesUtil.js [hunk 6]
```diff
@@ -19,7 +19,9 @@ const StudyListRouting = asyncComponent(() =>
   )
 );
 const StandaloneRouting = asyncComponent(() =>
-  import(/* webpackChunkName: "ConnectedStandaloneRouting" */ '../connectedComponents/ConnectedStandaloneRouting.js')
```
Comment (swederik) [inline platform/viewer/src/routes/routesUtil.js]: Unnecessary change you can remove from this PR. We can lint all files later

File: platform/viewer/public/config/default.js [hunk 7]
```diff
@@ -115,6 +115,11 @@ window.config = {
       label: 'W/L Preset 9',
       keys: ['9'],
     },
+    {
+      commandName: 'downloadAndZipSeriesOnActiveViewport',
```
Comment (swederik) [inline platform/viewer/public/config/default.js]: We do not want this enabled by default, you can take this out of the config.

**Comment by dannyrb** [comment]
> Can we use the DICOM-formatted names for variables for consistency around our codebase (StudyInstanceUID, etc..). James just made a huge PR to keep this consistent in the rest of the codebase.

This is why we need a linter.

File: extensions/dicom-p10-downloader/src/index.js (hunk 5 above)
Comment (emfol) [reply extensions/dicom-p10-downloader/src/index.js]: Actually, this is just a "default" DICOM Web Client instance built based on startup config which is only used if the used does not provide its own DICOMWebClient instance. Also, the command from extensions/dicom-p10-downloader/src/commandsModule.js file (line #11, listed above) uses the information inside the servers store to find out information about the currently active server and builds a DICOM Web Client from there.

File: platform/viewer/public/config/default.js (hunk 7 above)
Comment (emfol) [reply platform/viewer/public/config/default.js]: OK. I'll remove it in the following PR update.

File: platform/viewer/src/routes/routesUtil.js (hunk 6 above)
Comment (emfol) [reply platform/viewer/src/routes/routesUtil.js]: Yeah this freaking automatic linter keeps surprising me... I'll remove it. Thanks for that.

File: extensions/dicom-p10-downloader/src/commandsModule.js (hunk 3 above)
Comment (emfol) [reply extensions/dicom-p10-downloader/src/commandsModule.js]: Nope. This is a Blob URL that directly initiates the download and leaves the browsing session as is. But this was temporary and I have already changed this in order to give the file a name.

File: extensions/dicom-p10-downloader/src/commandsModule.js (hunk 2 above)
Comment (emfol) [reply extensions/dicom-p10-downloader/src/commandsModule.js]: Great point! I'll remove this conditional as this erroneous condition is properly handled (an exception is thrown) in thee lower layer.

File: extensions/dicom-p10-downloader/README.md (hunk 1 above)
Comment (emfol) [reply extensions/dicom-p10-downloader/README.md]: copy and paste hahaha

File: platform/viewer/public/config/default.js [hunk 8]
```diff
@@ -115,6 +115,11 @@ window.config = {
       label: 'W/L Preset 9',
       keys: ['9'],
     },
+    {
+      commandName: 'downloadAndZipSeriesOnViewports',
```
Comment (emfol) [inline platform/viewer/public/config/default.js]: @swederik, I left it here (along with the import extension on the "platform/viewer/src/index.js" file) in order to make the feature testable by the client. My plan is to remove it after the feature is approved and before it hits the master branch. Please let me know if you don't think it's necessary.

File: extensions/dicom-p10-downloader/src/commandsModule.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  save,
+  getDicomWebClientFromContext,
+  getSOPInstanceReferenceFromActiveViewport,
+  getSOPInstanceReferencesFromViewports,
+} from './utils';
+import downloadAndZip from './downloadAndZip';
+
+export function getCommands(context) {
+  const actions = {
+    // Example for running this command using Commands Manager
```
Comment (JamesAPetts) [inline extensions/dicom-p10-downloader/src/commandsModule.js]: You can add examples using jsdoc using the `@example` tag, like [here](https://github.com/cornerstonejs/cornerstoneTools/blob/e38e2f5c81b0f83cca5a29a03334bee287028a8a/src/store/setToolMode.js#L23).

File: extensions/dicom-p10-downloader/src/commandsModule.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    //       progress(status) {
+    //         console.info('Progress:', (status.progress * 100).toFixed(2) + '%');
+    //       }
+    //     }
+    //   },
+    //   'VIEWER'
+    // );
+    downloadAndZip({ servers, dicomWebClient, listOfUIDs, options }) {
+      return save(
+        downloadAndZip(
```
Comment (JamesAPetts) [inline extensions/dicom-p10-downloader/src/commandsModule.js]: Can you import this as an alias? Otherwise this function is hard to follow and looks like it is recursive.

File: extensions/dicom-p10-downloader/src/utils.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+      list,
+      validDicomUid(StudyInstanceUID),
+      validDicomUid(SeriesInstanceUID),
+      validDicomUid(SOPInstanceUID)
+    );
+  });
+  return list;
+}
+
+function save(thenable, listOfUIDs) {
```
Comment (JamesAPetts) [inline extensions/dicom-p10-downloader/src/utils.js]: `thenable` => `promise` ?

**Comment by dannyrb** [comment]
@jamesapetts, I think we were waiting on feedback for UI and the level of polish wanted before moving forward.

I'm also hesitant to put in any restrictions. I think those could/would be done by the UI. A future performance enhancement might push the zipping to a thread.

File: extensions/dicom-p10-downloader/src/commandsModule.js (hunk 9 above)
Comment (emfol) [reply extensions/dicom-p10-downloader/src/commandsModule.js]: Nice! I'll do that! Thank you!

File: extensions/dicom-p10-downloader/src/commandsModule.js (hunk 10 above)
Comment (emfol) [reply extensions/dicom-p10-downloader/src/commandsModule.js]: I see... Yes, I can do that.

File: extensions/dicom-p10-downloader/src/utils.js (hunk 11 above)
Comment (emfol) [reply extensions/dicom-p10-downloader/src/utils.js]: In this case I use "thenable" to differentiate between promise-like objects and actual Promise objects. This method is capable of handling both (actual) promises and promise-like objects (i.e., a thenable) and returns an actual Promise instance.

Comment (emfol) [reply extensions/dicom-p10-downloader/src/utils.js]: Oops! Indeed... In this context, promise is better! I'll change it.

**Comment by emfol** [comment]
@JamesAPetts, I pushed a few changes to prevent concurrent downloads and I could no longer reproduce the issue you reported (btw, you can track download progress from messages in console which, in the future, can be easily migrated to some sort of UI). Can you please try again? If you're able to reproduce it, please send me the link to the exam you tested with and as much information as you can about the issue.

**Comment by JamesAPetts** [comment]
Hey @emfol , all looks good to me.

**Comment by sebastianwoinar** [comment]
@JamesAPetts what happened to your extension? Any chance, that it can still be used with OHIF Viewer > 3.8?
